File: backend/src/main/java/com/picsou/mcp/Scopes.java
package com.picsou.mcp;

import java.util.Set;

/**
 * The complete vocabulary of access-key scopes ({@code domain:action}).
 *
 * <p>An access-key carries a subset of {@link #ALL}; each MCP {@code @Tool} method is gated by
 * exactly one of these via {@code @RequiresScope}. {@link #ALL} is the security allowlist —
 * {@code AccessKeyController} rejects any requested scope outside it, so a key can never be
 * granted a permission that no tool honours.
 *
 * <p>Read scopes end in {@code :read}; everything else mutates ({@code :write} / {@code :trigger}).
 * There is intentionally NO scope for credentials, MFA, admin settings, member management, or
 * data export — those operations have no tool, so no scope can reach them.
 */
public final class Scopes {

    private Scopes() {}

    // ── Read ────────────────────────────────────────────────────────────────
    public static final String ACCOUNTS_READ = "accounts:read";
    public static final String TRANSACTIONS_READ = "transactions:read";
    public static final String GOALS_READ = "goals:read";
    public static final String DASHBOARD_READ = "dashboard:read";
    public static final String PRICES_READ = "prices:read";
    public static final String FAMILY_READ = "family:read";

    public static final String BUDGET_CATEGORIES_READ = "budget:categories-read";
    public static final String BUDGET_RULES_READ = "budget:rules-read";
    public static final String BUDGET_TRANSACTIONS_READ = "budget:transactions-read";
    public static final String BUDGET_RECURRING_READ = "budget:recurring-read";
    public static final String BUDGET_ENVELOPES_READ = "budget:envelopes-read";
    public static final String BUDGET_DASHBOARD_READ = "budget:dashboard-read";
    public static final String OAUTH2_DISCOVER = "oauth2:discover";
    public static final String OAUTH2_SESSION_STATUS = "oauth2:session-status";

    // ── Write / trigger ───────────────────────────────────────────────────────
    public static final String ACCOUNTS_WRITE = "accounts:write";
    public static final String TRANSACTIONS_WRITE = "transactions:write";
    public static final String GOALS_WRITE = "goals:write";
    public static final String SYNC_TRIGGER = "sync:trigger";

    public static final String BUDGET_CATEGORIES_WRITE = "budget:categories-write";
    public static final String BUDGET_RULES_WRITE = "budget:rules-write";
    public static final String BUDGET_TRANSACTIONS_WRITE = "budget:transactions-write";
    public static final String BUDGET_ENVELOPES_WRITE = "budget:envelopes-write";

    /** Immutable allowlist of every valid scope. Used to validate key-creation requests. */
    public static final Set<String> ALL = Set.of(
        ACCOUNTS_READ, TRANSACTIONS_READ, GOALS_READ, DASHBOARD_READ, PRICES_READ, FAMILY_READ,
        BUDGET_CATEGORIES_READ, BUDGET_RULES_READ, BUDGET_TRANSACTIONS_READ, BUDGET_RECURRING_READ,
        BUDGET_ENVELOPES_READ, BUDGET_DASHBOARD_READ, OAUTH2_DISCOVER, OAUTH2_SESSION_STATUS,
        ACCOUNTS_WRITE, TRANSACTIONS_WRITE, GOALS_WRITE, SYNC_TRIGGER,
        BUDGET_CATEGORIES_WRITE, BUDGET_RULES_WRITE, BUDGET_TRANSACTIONS_WRITE, BUDGET_ENVELOPES_WRITE
    );
}

File: frontend/src/features/accessKeys/scopes.ts
/**
 * The access-key scope vocabulary, mirroring the backend allowlist
 * `com.picsou.mcp.Scopes#ALL`. The UI must offer exactly these — a scope the
 * backend doesn't recognise would make key creation fail with HTTP 400.
 *
 * Most read scopes end in `:read`; `budget:*` reads end in `-read` instead
 * (e.g. `budget:categories-read`), and the two `oauth2:*` introspection
 * scopes carry no read suffix at all — see `scopeGroup()`.
 */
export const ALL_SCOPES = [
  'accounts:read',
  'transactions:read',
  'goals:read',
  'dashboard:read',
  'prices:read',
  'family:read',
  'budget:categories-read',
  'budget:rules-read',
  'budget:transactions-read',
  'budget:recurring-read',
  'budget:envelopes-read',
  'budget:dashboard-read',
  'oauth2:discover',
  'oauth2:session-status',
  'accounts:write',
  'transactions:write',
  'goals:write',
  'sync:trigger',
  'budget:categories-write',
  'budget:rules-write',
  'budget:transactions-write',
  'budget:envelopes-write',
] as const

export type Scope = (typeof ALL_SCOPES)[number]

export type ScopeGroup = 'read' | 'write'

/** Read-only scopes that carry neither a `:read` nor a `-read` suffix. */
const READ_ONLY_OVERRIDES = new Set<string>(['oauth2:discover', 'oauth2:session-status'])

/**
 * Classifies a scope: `:read`/`-read` suffix, or a listed read-only override, → read;
 * everything else (`:write` / `:trigger`) → write.
 */
export function scopeGroup(scope: string): ScopeGroup {
  return scope.endsWith(':read') || scope.endsWith('-read') || READ_ONLY_OVERRIDES.has(scope)
    ? 'read'
    : 'write'
}

/** i18n-safe key for a scope: `accounts:read` → `accounts_read`. */
export function scopeI18nKey(scope: string): string {
  return scope.replace(':', '_')
}

export const READ_SCOPES: Scope[] = ALL_SCOPES.filter((s) => scopeGroup(s) === 'read')
export const WRITE_SCOPES: Scope[] = ALL_SCOPES.filter((s) => scopeGroup(s) === 'write')
